File: packages/client/src/pipeline/retry.ts
import { EcorpinError, type ActionMetadata } from "@ecorpin/core";

const SAFE_METHODS = new Set(["GET"]);

/**
 * An error is retried only if the taxonomy marks it `retryable` *and* the
 * action is either a safe HTTP method (GET) or explicitly `idempotent`
 * (architecture doc §16.5) — never on an ambiguous mutating outcome.
 */
export function isRetryable(err: unknown, action: ActionMetadata): boolean {
  if (!(err instanceof EcorpinError) || !err.retryable) return false;
  return SAFE_METHODS.has(action.method) || Boolean(action.idempotent);
}

function computeBackoffMs(attemptNumber: number): number {
  const base = 200 * 2 ** (attemptNumber - 1);
  const jitter = Math.random() * base * 0.3;
  return Math.min(base + jitter, 5000);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

/**
 * Runs `attempt` up to `maxRetries + 1` times total, backing off with
 * jitter between attempts, stopping as soon as `shouldRetry` returns false
 * for the error just thrown (architecture doc §7 failure branch).
 */
export async function withRetry<T>(
  attempt: (attemptNumber: number) => Promise<T>,
  maxRetries: number,
  shouldRetry: (err: unknown) => boolean
): Promise<T> {
  let lastError: unknown;
  for (let attemptNumber = 1; attemptNumber <= maxRetries + 1; attemptNumber += 1) {
    try {
      return await attempt(attemptNumber);
    } catch (err) {
      lastError = err;
      const isLastAttempt = attemptNumber > maxRetries;
      if (isLastAttempt || !shouldRetry(err)) {
        throw err;
      }
      await sleep(computeBackoffMs(attemptNumber));
    }
  }
  // Unreachable (the loop always returns or throws), kept for type-narrowing.
  throw lastError;
}

File: packages/client-py/src/ecorpin_client/retry.py
"""Retry-with-backoff, mirroring @ecorpin/client's pipeline/retry.ts."""
from __future__ import annotations

import random
import time
from typing import Callable, TypeVar

from .errors import EcorpinError

_SAFE_METHODS = {"GET"}

T = TypeVar("T")


def is_retryable(err: BaseException, action: dict) -> bool:
    """
    An error is retried only if the taxonomy marks it `retryable` *and* the
    action is either a safe HTTP method (GET) or explicitly `idempotent` —
    never on an ambiguous mutating outcome.
    """
    if not isinstance(err, EcorpinError) or not err.retryable:
        return False
    return action.get("method") in _SAFE_METHODS or bool(action.get("idempotent"))


def _compute_backoff_seconds(attempt_number: int) -> float:
    base = 0.2 * (2 ** (attempt_number - 1))
    jitter = random.random() * base * 0.3
    return min(base + jitter, 5.0)


def with_retry(
    attempt: Callable[[int], T],
    max_retries: int,
    should_retry: Callable[[BaseException], bool],
) -> T:
    """
    Runs `attempt` up to `max_retries + 1` times total, backing off with
    jitter between attempts, stopping as soon as `should_retry` returns
    False for the error just raised.
    """
    last_error: BaseException | None = None
    for attempt_number in range(1, max_retries + 2):
        try:
            return attempt(attempt_number)
        except Exception as err:  # noqa: BLE001 - re-raised below when not retried
            last_error = err
            is_last_attempt = attempt_number > max_retries
            if is_last_attempt or not should_retry(err):
                raise
            time.sleep(_compute_backoff_seconds(attempt_number))
    assert last_error is not None  # pragma: no cover - unreachable, loop always returns/raises
    raise last_error
